pip/qsharp/interop/qasm/_compile.py

```diff
@@ -0,0 +1,102 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT License.
+
+import warnings
+from time import monotonic
+from typing import Any, Callable, Dict, Optional, Union
+from ..._fs import read_file, list_directory, resolve
+from ..._http import fetch_github
+
+from ..._native import (  # type: ignore
+    compile_qasm_program_to_qir,
+)
+from ..._qsharp import QirInputData, get_interpreter, ipython_helper, TargetProfile
+from ... import telemetry_events
+
+
+def compile(
+    source: Union[str, Callable],
+    *args,
+    **kwargs: Optional[Dict[str, Any]],
+) -> QirInputData:
+    """
+    Compiles the OpenQASM source code into a program that can be submitted to a
+    target as QIR (Quantum Intermediate Representation).
+    Either a full program or a callable with arguments must be provided.
+
+    Args:
+        source (str): An OpenQASM program. Alternatively, a callable can be provided,
+            which must be an already imported global callable.
+        *args: The arguments to pass to the callable, if one is provided.
+        **kwargs: Additional keyword arguments to pass to the compilation when source program is provided.
+          - name (str): The name of the circuit. This is used as the entry point for the program.
+          - target_profile (TargetProfile): The target profile to use for code generation.
+          - search_path (Optional[str]): The optional search path for resolving file references.
+          - output_semantics (OutputSemantics, optional): The output semantics for the compilation.
+          - program_ty (ProgramType, optional): The type of program compilation to perform.
```

Comment:
Why abbreviate this to `program_ty` as opposed to `program_type`? That's a very Rust-y/compiler developer abbreviation and one I'm not sure our quantum audience would intuit.